packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+    event Redeemed(uint256 indexed tokenId, bytes32 indexed redemptionId, uint256 expiryTime);
+
+    /**
+     * @notice Constructor.
+     * @param _liveness liveness period between submission and verification of a reward.
+     * @param _bondToken ERC20 token that the bond is paid in.
+     * @param _bond size of the bond.
+     * @param _identifier identifier that should be passed to the optimistic oracle on dispute.
+     * @param _customAncillaryData custom ancillary data that should be sent to the optimistic oracle on dispute.
+     * @param _finder finder to look up UMA contract addresses.
+     */
+    constructor(
+        uint256 _liveness,
+        IERC20 _bondToken,
+        uint256 _bond,
+        bytes32 _identifier,
+        bytes memory _customAncillaryData,
+        FinderInterface _finder
+    ) {
+        require(liveness > 0, "liveness == 0");
+        liveness = _liveness;
+        finder = _finder;
+        require(_getCollateralWhitelist().isOnWhitelist(address(_bondToken)), "bond token not supported");
+        bondToken = _bondToken;
+        bond = _bond;
+        require(_getIdentifierWhitelist().isIdentifierSupported(_identifier), "identifier not supported");
+        identifier = _identifier;
+        SkinnyOptimisticOracleInterface skinnyOptimisticOracle = _getOptimisticOracle();
+        require(
+            skinnyOptimisticOracle.stampAncillaryData(_customAncillaryData, address(this)).length <=
+                skinnyOptimisticOracle.ancillaryBytesLimit(),
+            "ancillary data too long"
+        );
+        customAncillaryData = _customAncillaryData;
+    }
+
+    /**
+     * @notice Allows anyone to deposit reward tokens into the contract. Presumably, this would be the deployer or
+     * protocol that wishes to reward the users interacting with the system.
+     * @dev Once tokens are deposited, they cannot be withdrawn without claiming a reward. If a deployer wants an
+     * "escape hatch", they can create a special tokenId for this purpose.
+     * @param token ERC20 token that is being deposited.
+     * @param amount amount of rewards to deposit.
+     */
+    function depositRewards(IERC20 token, uint256 amount) public nonReentrant {
+        token.safeTransferFrom(msg.sender, address(this), amount);
+        emit Deposited(msg.sender, token, amount);
+    }
+
+    /**
+     * @notice Allows the caller to mint a token to the receiver and include a reward-relevant update event with it.
+     * This is intended to be used when the user first interacts with a reward-granting protocol.
+     * @dev if the user prefers to only mint a new token, they should call the mintNextToken function.
+     * @param receiver user that will receive the newly minted token.
+     * @param data arbitrary caller-generated data that will be associated with this update.
+     */
+    function mint(address receiver, bytes memory data) public nonReentrant returns (uint256 tokenId) {
```

Comment:
Do we need to do any sanitizing of the `data`? I'm curious what an example case of the data would be

Reply:
Right now, I've primarily been using `msg.data` which is just the abi encoded function that calls this one.

How would you suggest sanitizing it? Like turning it to utf8 or something?

Reply:
If the `data` is emitted by `UpdateTokens`, which is potentially used to determine how many rewards the NFT can be redeemed for, isn't it weird that the caller can set their own `data`? Perhaps i'm confusing its intended use case

Reply:
I see the confusion. This is why we log the caller. You can imagine an identifier that says to ignore all events where the caller isn't X. The idea is that there is a lot of flexibility for the users to tell this contract to emit any sort of data from this event from any sort of caller. Then offchain, the identifier can specify which callers are valid and how to decode the data they pass in.

Reply:
I'm not 100% sure if we should even have these events or just have the user to use their own contract to associate events with the tokenIds. But it seemed fairly simple to add without any real downside.

The one use case that this could enable is a privileged EOA (potentially specified in the contract's customAncillaryData) could call this method to add rewards to individual tokens without needing to use or create a contract at all.

Reply:
This makes sense